src/kernel/mpipeline.ml

```diff
@@ -252,35 +274,37 @@ let process
           raw_source, Some ast
     )) in
   let reader =
-    timed_lazy reader_time
-      (lazy
-        (let (lazy ((_, pp_result) as source)) = source in
-          let config = Mconfig.normalize config in
-          Mocaml.setup_reader_config config;
-          let cache_disabling =
-            match (config.merlin.use_ppx_cache, pp_result) with
-            | false, _ -> Some "configuration"
-            | true, Some _ ->
-              (* The cache could be refined in the future to also act on the
-                 PP phase. For now, let's disable the whole cache when there's
-                 a PP. *)
-              Some "source preprocessor usage"
-            | true, None -> None
-          in
-          let { Reader_with_cache.output = { result; cache_version }; _ } =
-            Reader_with_cache.apply ~cache_disabling
-              { source; for_completion; config }
-          in
-          let cache_version =
-            if Option.is_some cache_disabling then None else Some cache_version
-          in
-          { Reader.result; config; cache_version }
-    )) in
-  let ppx = timed_lazy ppx_time (lazy (
+    Reader.cache_stat_lazy reader_cache_hit
+      (timed_lazy reader_time
+        (lazy
+          (let (lazy ((_, pp_result) as source)) = source in
+            let config = Mconfig.normalize config in
+            Mocaml.setup_reader_config config;
+            let cache_disabling =
+              match (config.merlin.use_ppx_cache, pp_result) with
+              | false, _ -> Some "configuration"
+              | true, Some _ ->
+                (* The cache could be refined in the future to also act on the
+                   PP phase. For now, let's disable the whole cache when there's
+                   a PP. *)
+                Some "source preprocessor usage"
+              | true, None -> None
+            in
+            let { Reader_with_cache.output = { result; cache_version }; cache_was_hit } =
+              Reader_with_cache.apply ~cache_disabling
+                { source; for_completion; config }
+            in
+            let cache_version =
+              if Option.is_some cache_disabling then None else Some cache_version
+            in
+            { Reader.result; config; cache_version; cache_was_hit }
```

Comment:
Like here we could update the reference and not return "cache_was_it".

Or I am missing some subtelty ?

Reply:
Nope, it was just a sketch. Now I made it proper according to your suggestion